### ISSUE TITLE
Use annotations from libraries without prefix

### ISSUE THREAD

**Issue by oberlies** [raml-org/raml-spec#207]
The "uses" property is useful for splitting a larger RAML file into smaller, manageable parts. For example, I would like to split out our numerous annotations into a separate RAML file/library. With the current definition of "uses", this then forces me to use the annotations with a prefix. This may make sense when using third party annotation libraries, but is unnecessary bloat for our own annotations.

So, it should be possible to import a set of annotations from a separate file and still use them without prefix.

Example:
* annotations.raml:

        #%RAML 1.0 Library
        
        annotationTypes:
          errors:
            allowedTargets: API
            parameters:
              value: errorMetaType[]
          errorResponses:
            allowedTargets: Method
            parameters:
              codes: string[]
        types:
          errorMetaType:
            type: object
            properties:
              httpStatus: string
              description: string

* api.raml (current syntax):

        #%RAML 1.0
        title: ""
        uses:
          - annotations: !include annotations.raml
        (annotations.errors):
          value:
            invalidRequestBody:
              httpStatus: badRequest (400)
              description: The request body is not a JSON object
        /resource:
          post:
            description: Creates a new resource
            responses:
              201:
                description: ...
            (annotations.errorResponses):
              codes: [invalidRequestBody]

It should be possible to use `(errors)` and `(errorResponses)` in api.raml. Currently this doesn't work.

**Comment by sichvoge** [comment]
IMHO, that is not a good idea. Having that possibility everyone would not use the prefixes anymore as it is a shortcut. We know how much we love shortcuts ;) That will definitely complex the discovery process for humans, but also machines of where these fragments are coming from. Furthermore, how do you distinct between two, for example, annotation types that has been described in two different files? In a large scale environment that can end up quite messy.

Maybe the way is not very much "comfortable", but it's very good to distinguish between different libs containing different fragments where another with the same name might be defined across libraries as well.

**Comment by oberlies** [comment]
> That will definitely complex the discovery process for humans, but also machines of where these fragments are coming from

This is true, but IMHO prefixes are a poor choice to solve this problem (see also #204). Instead, it should be possible to import symbols from other files and use them either with a qualified or unqualified name. Just like e.g. imports in Java.

**Comment by oberlies** [comment]
By mis-using the extensions concept, I can use the annotations from a separate file without prefix. For the example, this works in the following way:
* Make annotations.raml a normal RAML file (`#%RAML 1.0`)
* Change the main file api.raml to an extension (`#%RAML 1.0 Extension`)
* Add `masterRef: annotations.raml` in api.raml

This workaround is not nice, but better than having to qualify all annotations.

**Comment by sichvoge** [comment]
> This is true, but IMHO prefixes are a poor choice to solve this problem (see also #204). Instead, it should be possible to import symbols from other files and use them either with a qualified or unqualified name. Just like e.g. imports in Java.

That is true. That also would mean the parser and spec needs to define how we handle collisions; two types that are syntactically the same in two different libraries. Right?

**Comment by gardea** [comment]
On a personal note, I initially did not appreciate *!include*, if only because the "reference" implementation of the parser was not (and still isn't, apparently) capable of pointing out syntax issues correctly (i.e. the context of the file is lost in the merge). I thought then that include/import in XSD was so much better executed. These days, I grew to like seeing people not getting lost in the semantics of it; it simply adds, in an unambiguous way, to the symbol space. I think uses may be a step in the right direction; as long as we don't reintroduce "namespace", etc. which proved to be such a misunderstood concept in XML (at least). I would really try to see if problems in scope for RAML could be addressed without namespaces. uses brings it's own issues, still undocumented, which would have to explain how to ensure files are not considered twice in contributing to the symbol space (as it were with special XSD schemas, where well known schemas would point out to local copies of known namespaces, thus creating problems when known schemas were put together in yet another context).

**Comment by oberlies** [comment]
> That also would mean the parser and spec needs to define how we handle collisions; two types that are syntactically the same in two different libraries. Right?

Yes, I suppose this would need to be defined.
But primarily I would try to avoid collisions altogether: Instead of importing all symbols from a file, there could be a `uses` syntax that explicitly lists the symbols to be imported. And then we can simply disallow duplicate imports or import and inline declaration of the same symbol.

**Comment by gardea** [comment]
It may be impossible to avoid given the very reason it exists: one may use two different libraries, which in turn both point to a third one.

**Comment by sichvoge** [comment]
@oberlies & @gardea are you talking about something like this?

```yaml
uses:
  - !include lib1.raml#Person
  - !include lib3.raml#Annotation1

/resource:
  (Annotation1): something
  get:
    responses: 
      200:
        body: 
          application/json:
            type: Person
```

Just trying to continue our conversation by throwing in some examples.

**Comment by oberlies** [comment]
@sichvoge To my understanding yes. Having something like this would be a solution for my use case.

**Comment by gardea** [comment]
@sichvoge, I would think that `!include lib1.raml#Person` is rather a bad example. A library defines multiple symbol spaces, as defined by the key: types, traits, resourceTypes, etc. The above path causes ambiguity, is in should it be `!include lib1.raml#/types/Person` or maybe `!include lib1.raml#/traits/Person` , etc. ?

More so, would this "qualified" include imply that you're only bringing in that particular definition, and maybe all its dependencies? How would you define dependencies for a type system that allow for substitution groups?

I've started defining a schema for the RAML 1.0 spec (similar to how Swagger 2.0 has one) and I have to admit, this spec puzzles me. While I appreciate the benefits I see being introduced, I am not sure it is the most efficient way to achieve it (includes vs. libraries vs. overlays vs. extensions). I am looking at this as someone that also needs to train people, and so allow the process of authoring specs a natural and fluent way, rather than full of twists (a.k.a. decision points).

**Comment by petrochenko-pavel-a** [comment]
Hi guys I am thinking about a big more consistent (with current approach) syntax for importing types in the current namespace `~` - means use the same name as in original library otherwise you can customise another name.


```raml
uses:
  ~: !include lib1.raml#Person
  AAA: !include lib3.raml#Annotation1

/resource:
  (Annotation1): something
  get:
    responses: 
      200:
        body: 
          application/json:
            type: Person
``

>More so, would this "qualified" include imply that you're only bringing in that particular definition, and >maybe all its dependencies?
As for me it always means bringing particular definition.

>!include lib1.raml#/types/Person
yes there is some ambiguity, I still like a shorter syntax however, I understood its problems.

>How would you define dependencies for a type system that allow for substitution groups?
May you please elaborate on it a little bit.

>I've started defining a schema for the RAML 1.0 spec 
This is an interesting topic you want we can collaborate on it.

Regards,
Pavel

**Comment by sichvoge** [comment]
Additionally to that you could even include a whole library using the same approach.

```yaml
uses:
  ~: !include lib1.raml
```

and use their respective names, right @petrochenko-pavel-a?
